packages/wallet/src/services/AddressDiscovery/HDSequentialDiscovery.ts

```diff
@@ -0,0 +1,86 @@
+ * and the algorithm will still continue with discovery if the address was previously used.
+ *
+ * If the wallet hits gap limit of unused addresses in a row, it expects there are
+ * no used addresses beyond this point and stops searching the address chain.
+ */
+export class HDSequentialDiscovery implements AddressDiscovery {
+  readonly #stakeKeyIndexLimit: number;
+
+  constructor(stakeKeyIndexLimit: number = DEFAULT_STAKE_KEY_INDEX_LIMIT) {
+    this.#stakeKeyIndexLimit = stakeKeyIndexLimit;
+  }
+
+  public async discover(
+    asyncKeyAgent: AsyncKeyAgent,
+    chainHistoryProvider: ChainHistoryProvider,
+    lookAheadCount: number
+  ): Promise<GroupedAddress[]> {
+    let currentGap = 0;
+    let currentIndex = 0;
+    const addresses = new Set<GroupedAddress>();
+    // The AsyncKeyAgent emits every time it derives an address; however; we don't want to emit all the addresses
+    // that are derived during the discovery process as some of those are gap addresses that we want to discard, so,
+    // we must use the inner key agent which doesn't emit. All discovered addresses will be emitted at the end of the
+    // process by `asyncKeyAgent.setKnownAddresses`
+    const keyAgent = await asyncKeyAgent.getKeyAgent();
+
+    // Add to our known address pool, all possible stake keys combined with payment credential at index 0.
+    for (let index = 0; index < this.#stakeKeyIndexLimit; ++index) {
+      const address = await keyAgent.deriveAddress({ index: 0, type: AddressType.External }, index);
+      addresses.add(address);
+    }
+
+    // Search for all base addresses composed with the stake key at index 0. We are assuming this is
+    // the use case with multi address wallets.
+    while (currentGap <= lookAheadCount) {
+      const address = await keyAgent.deriveAddress({ index: currentIndex, type: AddressType.External }, 0);
+
+      // We could fetch transactions from multiple addresses in a single query and then parse/organize the results, however, if these addresses
+      // contain huge transaction history we may get stuck retrieving all the transactions for longer than we can just query each address individually.
+      const txs = await chainHistoryProvider.transactionsByAddresses({
+        addresses: [address.address],
+        pagination: {
+          limit: 1,
+          startAt: 0
+        }
+      });
+
+      if (txs.totalResultCount > 0) {
+        currentGap = 0;
+        addresses.add(address);
+      } else {
+        ++currentGap;
+      }
+
+      ++currentIndex;
+    }
+
+    const result = [...addresses.values()];
+
+    // We need to make sure the addresses are sorted since the wallet assumes that the first address
+    // in the list is the change address (payment cred 0 and stake cred 0).
+    result.sort((a, b) => a.index - b.index || a.stakeKeyDerivationPath!.index - b.stakeKeyDerivationPath!.index);
+    asyncKeyAgent.setKnownAddresses(result);
```

Comment:
`discover` should be an atomic operation that keeps the KeyAgent in it's original state if it fails at any points (which is very possible when making multiple http requests). Since `deriveAddress` mutates the key agent, you could save knownAddresses at the start of the function and rolling back to it in a `catch`. However see my other comment about refactoring deriveAddress

Reply:
KeyAgentBase updates [knownAddresses](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/key-management/src/KeyAgentBase.ts#L107) so it doesn't matter whether you're using async or non-async variant

Reply:
Would be great to add a test where the discovery fails, asserting that keyAgent.knownAddresses weren't changed